python/cudf/cudf/core/udf/strings_typing.py

```diff
@@ -0,0 +1,197 @@
+# Copyright (c) 2022, NVIDIA CORPORATION.
+
+import operator
+
+from numba import types
+from numba.core.typing import signature as nb_signature
+from numba.core.typing.templates import AbstractTemplate, AttributeTemplate
+from numba.cuda.cudadecl import registry as cuda_decl_registry
+
+from strings_udf._typing import StringView, size_type, string_view
+
+from cudf.core.udf import masked_typing
+from cudf.core.udf._ops import comparison_ops
+from cudf.core.udf.masked_typing import MaskedType
+
+masked_typing.MASKED_INIT_MAP[types.pyobject] = string_view
+masked_typing.MASKED_INIT_MAP[string_view] = string_view
+
+
+def _is_valid_string_arg(ty):
+    return (
+        isinstance(ty, MaskedType) and isinstance(ty.value_type, StringView)
+    ) or isinstance(ty, types.StringLiteral)
+
+
+# String functions
+@cuda_decl_registry.register_global(len)
+class MaskedStringViewLength(AbstractTemplate):
+    """
+    provide the length of a cudf::string_view like struct
+    """
+
+    def generic(self, args, kws):
+        if isinstance(args[0], MaskedType) and isinstance(
+            args[0].value_type, StringView
+        ):
+            return nb_signature(MaskedType(size_type), args[0])
+
+
+@cuda_decl_registry.register_global(operator.contains)
+class MaskedStringViewContains(AbstractTemplate):
+    """
+    return a boolean indicating if a substring is found in a string
+    """
+
+    def generic(self, args, kws):
+        if _is_valid_string_arg(args[0]) and _is_valid_string_arg(args[1]):
+            return nb_signature(
+                MaskedType(types.boolean),
+                MaskedType(string_view),
+                MaskedType(string_view),
+            )
+
+
+class MaskedStringViewCmpOp(AbstractTemplate):
+    """
+    return the boolean result of `cmpop` between to strings
+    since the typing is the same for every comparison operator,
+    we can reuse this class for all of them.
+    """
+
+    def generic(self, args, kws):
+        if _is_valid_string_arg(args[0]) and _is_valid_string_arg(args[1]):
+            return nb_signature(
+                MaskedType(types.boolean),
+                MaskedType(string_view),
+                MaskedType(string_view),
+            )
+
+
+for op in comparison_ops:
+    cuda_decl_registry.register_global(op)(MaskedStringViewCmpOp)
+
+
+@cuda_decl_registry.register_global(len)
+class StringLiteralLength(AbstractTemplate):
```

Comment:
It won't matter from a technical perspective. The `generic` function should get searched by numba no matter what, and if it emits a signature then numba says 👍 . I had originally added this separately because in my mind, the statements `len(x)` when `x` is a `MaskedType` (e.g. a maybe valid string obtained by indexing into a source column via []) and the python statement `len('abc')` which is a string injected as a literal. 

It may even be that this overload is not needed. Sometimes we can just add a cast that promotes a literal to a `MaskedType`, like `abc -> MaskedType(abc, valid=True)` and then only retain the overload on `MaskedType`, and numba may promote it automatically. I can experiment with this.

Reply:
I'm fine with whatever you decide here.